### PR TITLE
feat(metrics): add Autoware-compatible EPDMS proxy validation

### LINKED ISSUE #142: Adopt EPDMS (Autoware planning_data_analyzer / NAVSIM) as a shared validation metric
## Summary

Bring the **faithful EPDMS** (Autoware `planning_data_analyzer` / NAVSIM v2 spec) that already exists in another repositry into Diffusion-Planner's validation as a shared, consistent metric — instead of each repo growing its own divergent "EPDMS-style" implementation.

This is **Track B**, split out from #130. The two are complementary:

- **#130 (Track A)** — surface Diffusion-Planner's *reward* subscores (`rlvr/reward.py` → `diffusion_planner.metrics`) in validation so model selection matches **what RLVR optimizes**. These are reward-shaping subscores (custom thresholds, goal-based EP, penalty signs) — **EPDMS-inspired, not faithful EPDMS**.
- **This issue (Track B)** — adopt the **faithful** EPDMS as a validation / closed-loop-proxy metric, aligned with the official Autoware/NAVSIM definition

## What already exists
- `pdms_navsim.py` — formula-faithful NAVSIM PDMS/EPDMS port (numpy / scipy / shapely): `no_at_fault_collision`, `time_to_collision`, `ego_progress(_with_gate)`, `comfort_score`, `extended_comfort`, `lane_keeping_score`, `dac_from_road_borders`, `ddc_from_route_lanes`, `aggregate_pdms`, … weights from NAVSIM `PDMScorerConfig` (EP 5, TTC 5, LK 2, EC 2, comfort 2).
- `pdms_proxy.py` — torch-facing wrapper (`pdms_proxy`, `pdms_proxy_masked`, `pdms_proxy_modes`, `epdms_human_filtered`); metrics in `[0, 1]`, higher = better.
- `val_loop.py` — already logs `val/pdms_proxy`, `val/no_collision`, `val/ttc`, `val/dac`, `val/lane_keeping`, `val/extended_comfort` and the human-filtered aggregate.

## How it differs from Diffusion-Planner's metrics

| | DP `diffusion_planner.metrics` (reward subscores) | EPDMS |
|---|---|---|
| backend | batched torch (GPU) | numpy / scipy / shapely (CPU, per-sample) |
| fidelity | RLVR reward formulation | bit-exact NAVSIM / planning_data_analyzer |
| output | penalties / gates (mixed signs) | `[0, 1]`, higher = better |
| EP (progress) | goal-distance based (not human-referenced) | expert/GT-referenced (proposal-relative, 5 m gate) |
| human filter | — | `epdms_human_filtered` ✅ |
| extended comfort / oracle best-of-M | — | ✅ |
| TLC | red-light penalty | pending (defaults to 1.0) |

## Data / integration gaps for DP validation

Target: `diffusion_planner/diffusion_planner/validate_model.py` (validation core extracted in #133), additive only — existing `avg_loss_*` outputs unchanged.

- **agent boxes** — reads per-timestep `[N, 9]` ego-frame agent boxes from a `.gt.tar` sidecar; DP has `neighbor_agents_future` (+ per-box dims from `neighbor_agents_past`) and would need a converter to that format (the NC/TTC critical path).
- **map geometry** — extract road-border polylines, route polygons, lane/intersection rings from DP's `line_strings` / `lanes` / `route_lanes` / `polygons` tensors
- **new deps** — scipy + shapely.

## Out of scope

- #130 (Track A) — reward-subscore validation parity.
- Bit-exact closed-loop PDMS with a bicycle-model simulator + nuplan maps (separate effort).

### PR DESCRIPTION
**Pull request by Max-Bin**
## Summary

Fixes #142.

This PR adds an offline EPDMS **proxy validation** path for Diffusion-Planner tensors:

- ports NAVSIM / Autoware-compatible open-loop subscore helpers into `planner_metrics`
- adds a strict `synthetic_epdms` aggregation wrapper matching `autoware_planning_data_analyzer` C++ aggregation semantics
- logs validation metrics under `valid_epdms/*` when explicitly enabled
- keeps existing `avg_loss_*` validation outputs unchanged
- keeps EPDMS disabled by default so existing training and validation behavior is unchanged

Important: this PR is **not** a bit-exact replacement for the Autoware `planning_data_analyzer` C++ evaluator. The final aggregation logic is intentionally matched to C++, but several raw subscores are reconstructed from existing DP NPZ tensors and therefore are proxy values.

## What is C++-matched

The synthetic aggregation follows `autoware_planning_data_analyzer/src/metrics/epdms/aggregation`:

- multiplicative terms: `NC * DAC * DDC * TLC`
- weighted terms: `EP(5) + TTC(5) + LK(2) + HC(2) + EC(2)`
- fixed weighted denominator: `16`
- raw synthetic EPDMS is available only when every required raw subscore is available
- human-filtered synthetic EPDMS uses the same human-zero filter rule
- `extended_comfort` bypasses the human filter, matching the C++ implementation

Missing raw subscores are reported with `<metric>_available = 0` instead of being silently treated as neutral passes. Logged synthetic EPDMS should only be compared as a C++-compatible aggregate when the matching synthetic availability flag is true.

## Important limitations vs Autoware C++

This validation path first tries to reconstruct all feasible inputs from the current DP NPZ batch. However, the DP NPZs do not contain the full ROS, RouteHandler, Lanelet2, traffic-signal, and temporal planner-state inputs consumed by the C++ evaluator. Therefore the following raw subscores are not bit-exact C++ ports.

| Subscore | Current PR behavior | Why it differs from C++ | Why old DP NPZs cannot recover exact C++ behavior |
| --- | --- | --- | --- |
| `TLC` traffic-light compliance | Unavailable unless explicit `traffic_light_compliance_values` are supplied | C++ needs `TrafficLightGroupArray`, regulatory elements / stop lines, route-relevant lanelets, turn indicator status, and ego footprints | Existing DP NPZ tensors do not store signal states, stop-line regulatory-element IDs, traffic-light group association, or synchronized signal timing |
| `EC` extended comfort | Unavailable unless explicit `extended_comfort_values` are supplied | C++ compares the previous selected planner trajectory against the current selected planner trajectory using trajectory timestamps, overlap shift, acceleration, jerk, yaw-rate, and yaw-acceleration RMS thresholds | A single old NPZ sample stores GT/model horizon tensors, not the previous deployed/planner selected trajectory, its header timestamp, or its trajectory-point acceleration fields. Exact EC requires sequential planner outputs or generation-time storage |
| `HC` history comfort | Computed from the trajectory pose tensor with derived finite-difference kinematics | C++ combines past odometry/acceleration history with future `TrajectoryPoint` acceleration, then computes comfort signals with the C++ local polynomial filter | Old NPZs do not store synchronized odometry and acceleration-history messages. Pose-only tensors cannot reconstruct measured longitudinal/lateral acceleration exactly |
| `EP` ego progress | Uses prediction projected onto the GT future polyline as an offline reference path | C++ projects trajectory start/end onto route-relevant lanelets from `RouteHandler`, gates on multiplicative metric availability, and uses route arc coordinates | Old NPZs do not retain Lanelet2 IDs, full route handler state, lanelet map pointer, or exact route-relevant lanelet collection used by C++ |
| `DAC` drivable-area compliance | Uses DP `line_strings` road-border polylines and local lane-derived polygons | C++ evaluates ego footprints against semantic drivable areas from `RouteHandler` / Lanelet2, including road lanelets, shoulders, intersections, hatched road markings, parking lots, and road-border fallback | NPZ map tensors are cropped/vectorized local features and do not preserve full Lanelet2 semantic layers, lanelet IDs, topology, or all candidate polygons from the C++ search windows |
| `DDC` driving-direction compliance | Uses DP `route_lanes` polygons to approximate on-route / oncoming progress | C++ builds local route-consistent lanelets using route lanelets, same-direction nearby road/shoulder lanelets, intersection areas, yaw consistency, and Lanelet2 topology | Old NPZs do not store the full lanelet topology, lanelet IDs, same-direction lanelet query results, or intersection regulatory semantics needed for the C++ local context |
| `LK` lane keeping | Uses route centerline distance, intersection polygons, and simplified queue/lane-change handling | C++ evaluates `LaneKeepingEvaluationPoint`s with route lanelet IDs, route centerlines, intersection flags, speed/progress queue exemption, queue-release grace, and lane-change indicator windows | Old NPZs do not store future turn-indicator windows, lane-change windows, route lanelet IDs, or the exact C++ lane-keeping evaluation points |
| `NC` no-at-fault collision | Converts neighbor GT futures and current size estimates into per-timestep boxes | C++ uses logged object tracks / timed tracked objects, object IDs, classifications, interpolated object polygons, unknown-object filtering, object velocity, route/map area flags, and exact `VehicleInfo` footprint offsets | Old NPZs do not store raw tracked-object history, object IDs, full classifications, exact polygon shapes, interpolation timestamps, or full C++ `VehicleInfo` parameters |
| `TTC` time-to-collision | Uses neighbor GT future boxes and ego footprint projection | C++ uses timed tracked objects / object tracks, object classification, ahead/behind checks, route/map area flags, stopped-ego logic, and exact footprint projections at fixed future offsets | Old NPZs do not store raw timed tracked-object messages, exact object IDs/classes/shapes, full timestamped interpolation state, or the C++ map-area evaluation outputs |

## Map/data inputs used by this PR

This PR does not add an external Lanelet2 map DB or Autoware runtime map reader. The validation code derives the map-dependent proxy inputs from tensors already loaded by the existing DP NPZ dataloader:

- `line_strings`: converted to road-border polylines for a DAC proxy when `--epdms_eval_use_road_border` is enabled
- `route_lanes`: converted to route lane polygons for DDC proxy and route centerlines for LK proxy
- `lanes` and `polygons`: converted to local lane/intersection polygons for NC/TTC area flags
- `neighbor_agents_future`, `neighbor_agents_past`, and `neighbor_agents_future_valid`: converted to future agent boxes for NC/TTC proxy
- `ego_shape`: used as the ego footprint size for border/collision/TTC checks

## Why unavailable is intentional

For this PR, unavailable is safer than silently filling a missing C++ input with a neutral score. In particular:

- `TLC` cannot be reconstructed from trajectory/map tensors because traffic-signal state and regulatory-element association are not present.
- `EC` cannot be reconstructed exactly from independent old NPZ samples because it is a temporal consistency metric between consecutive selected planner trajectories.
- Map-heavy metrics can be approximated from cropped DP map tensors, but the cropped tensors are not equivalent to the C++ `RouteHandler` + Lanelet2 map queries.

Future data generation can make the metric closer to C++ by saving the missing C++ inputs directly, for example traffic-light regulatory context, full route/lanelet IDs, selected planner trajectory history for EC, exact `VehicleInfo`, object IDs/classes/polygons, and synchronized object/ego timestamps.

## Validation

- `python3 -m py_compile` on edited Python files
- import check with the existing training venv
- targeted synthetic EPDMS formula sanity check covering fixed denominator, multiplicative product, availability, and human filter

A pytest test file is included for the synthetic aggregation formula; the local training venv used here does not include `pytest`, so the equivalent formula sanity check was executed directly.

## What a full map side input would and would not fix

Providing a full Lanelet2 / Autoware map side input would improve only the map-query part of this proxy. It could make `EP`, `DAC`, `DDC`, and `LK` closer to C++ if each NPZ sample also has enough metadata to recover the C++ query frame, such as map name/version, global ego pose, route/lanelet IDs, route handler state, and timestamp alignment.

A full map alone would still not make the metric bit-exact because several required C++ inputs are not map data:

- `TLC` still needs synchronized traffic-light signal states, stop-line/regulatory-element association, and route-relevant lanelets.
- `EC` still needs the previous selected planner trajectory and current selected planner trajectory with timestamps and trajectory-point acceleration fields.
- `HC` still needs synchronized odometry and acceleration history, not only future pose tensors.
- `NC`/`TTC` still need raw tracked/timed objects with IDs, classifications, object polygons, velocities, and interpolation timestamps.

Therefore, the exact C++ path is not simply "load a full map in the dataloader". The robust path is either to run the Autoware C++ `planning_data_analyzer` evaluator on synchronized ROS/Autoware inputs, or to extend future data generation to save all evaluator inputs as sidecars.

OnePlanner follows the same broad compromise: it stores and consumes local vector-map tensors for model input / proxy metrics, not a full C++ RouteHandler/Lanelet2 evaluator state. Its scene tensors include bounded local features such as lanes, route lanes, polygons, line strings, traffic-light one-hot attributes, and speed-limit fields. That is useful for learning and proxy scoring, but it is still not equivalent to the full Autoware C++ EPDMS runtime context.

### CODE REVIEW

File: diffusion_planner/diffusion_planner/validate_model.py [hunk 1]
```diff
@@ -158,6 +346,18 @@ def validate_model(model, val_loader, args, return_pred=False) -> tuple[float, f
...
+                prediction,
+                ego_future,
+                neighbors_future,
+                neighbors_future_valid,
+                denorm_inputs,
+                args,
+            )
+            for key, val in epdms_dict.items():
+                total_result_dict[f"epdms_{key}"].append(val.detach().cpu())
+
```
Comment (SakodaShintaro) [inline diffusion_planner/diffusion_planner/validate_model.py:366]: I feel that it is convenient to be able to set these parameters in `valid_predictor.py`.

File: diffusion_planner/diffusion_planner/train_config.py [hunk 2]
```diff
@@ -97,6 +97,14 @@ class TrainConfig:
     neighbor_collision_margin_pedestrian: float = 1.0
     neighbor_collision_margin_bicycle: float = 0.5
 
+    # Validation-only Autoware-aligned EPDMS metrics. Default-off to keep
+    # existing training and validation behavior unchanged unless explicitly enabled.
+    enable_epdms_eval: bool = False
+    # Backward-compatible alias for local scripts that used PDMS naming.
+    enable_pdms_eval: bool = False
+    epdms_eval_use_agent_boxes: bool = True
+    epdms_eval_use_road_border: bool = True
```
Comment (go-sakayori) [inline diffusion_planner/diffusion_planner/train_config.py]: Dead code since the parameter is overwritten by argparse

Comment (Max-Bin) [reply diffusion_planner/diffusion_planner/train_config.py]: Addressed in https://github.com/tier4/Diffusion-Planner/pull/173/commits/ee9fa96a62516f3d13f5e91583950eed3c072722

File: diffusion_planner/diffusion_planner/validate_model.py (hunk 1 above)
Comment (Max-Bin) [reply diffusion_planner/diffusion_planner/validate_model.py:366]: Fixed. The valid_predictor.py now exposes the EPDMS-related validation options, and the aggregated epdms_* metrics are included in valid_dict.json when EPDMS evaluation is enabled.

**Review by SakodaShintaro — APPROVED** [review]

File: diffusion_planner/diffusion_planner/valid_config.py [hunk 3]
```diff
@@ -23,3 +23,8 @@ class ValidConfig:
     predicted_neighbor_num: int = 32
     ddp: bool = True
     port: str = "22323"
+
+    enable_epdms_eval: bool = False
```
Comment (SakodaShintaro) [inline diffusion_planner/diffusion_planner/valid_config.py]: I feel that it is also good to change the default value to `True`

Comment (Max-Bin) [reply diffusion_planner/diffusion_planner/valid_config.py]: @SakodaShintaro  Fixed in https://github.com/tier4/Diffusion-Planner/pull/173/commits/3d1ab6aa58d9f36c47e5e3eea14c29a460046d76
